diff --git a/core/modules/media/src/Plugin/media/Source/File.php b/core/modules/media/src/Plugin/media/Source/File.php
index 1c963ef9636b..15fdf7e4fd00 100644
--- a/core/modules/media/src/Plugin/media/Source/File.php
+++ b/core/modules/media/src/Plugin/media/Source/File.php
@@ -66,6 +66,8 @@ public function getMetadata(MediaInterface $media, $attribute_name) {
     }
     switch ($attribute_name) {
       case static::METADATA_ATTRIBUTE_NAME:
+        return $media->getName();
+
       case 'default_name':
         return $file->getFilename();
 
