diff --git a/src/Widget/WidgetPluginBase.php b/src/Widget/WidgetPluginBase.php
index 59053aec351e..5d671c1cf599 100644
--- a/src/Widget/WidgetPluginBase.php
+++ b/src/Widget/WidgetPluginBase.php
@@ -206,6 +206,7 @@ abstract class WidgetPluginBase extends PluginBase implements WidgetPluginInterf
     $items['#wrapper_attributes'] = ['class' => $classes];
     $items['#attributes']['data-drupal-facet-item-id'] = Html::getClass($this->facet->getUrlAlias() . '-' . strtr($result->getRawValue(), ' \'\"', '---'));
     $items['#attributes']['data-drupal-facet-item-value'] = $result->getRawValue();
+    $items['#attributes']['data-drupal-facet-item-count'] = $result->getCount();
     return $items;
   }
 
